flink-autoscaler/src/main/java/org/apache/flink/autoscaler/utils/MemoryTuningUtils.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.flink.autoscaler.metrics.EvaluatedScalingMetric;
+import org.apache.flink.autoscaler.metrics.ScalingMetric;
+import org.apache.flink.configuration.MemorySize;
+import org.apache.flink.configuration.TaskManagerOptions;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Optional;
+
+/** Tunes the TaskManager memory. */
+public class MemoryTuningUtils {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MemoryTuningUtils.class);
+
+    public static Optional<MemorySize> tuneTaskManagerHeapMemory(
+            JobAutoScalerContext<?> context,
+            EvaluatedMetrics evaluatedMetrics,
+            Map<JobVertexID, ScalingSummary> scalingSummaries) {
+
+        var config = context.getConfiguration();
+        if (!config.get(AutoScalerOptions.MEMORY_TUNING_ENABLED)) {
+            return Optional.empty();
+        }
+
+        var globalMetrics = evaluatedMetrics.getGlobalMetrics();
+        double avgHeapSize = globalMetrics.get(ScalingMetric.HEAP_AVERAGE_SIZE).getAverage();
+
+        double numTaskSlotsUsed = globalMetrics.get(ScalingMetric.NUM_TASK_SLOTS_USED).getCurrent();
+        int taskSlotsPerTm = config.get(TaskManagerOptions.NUM_TASK_SLOTS);
+        int currentNumTMs = (int) Math.ceil(numTaskSlotsUsed / taskSlotsPerTm);
+
+        double usedTotalHeapSize = currentNumTMs * avgHeapSize;
+        LOG.info("Total used heap size: {}", new MemorySize((long) usedTotalHeapSize));
+        usedTotalHeapSize *= computeDataChangeRate(evaluatedMetrics);
+        LOG.info("Resized total heap size: {}", new MemorySize((long) usedTotalHeapSize));
+
+        int numTaskSlotsAfterRescale =
+                ResourceCheckUtils.estimateNumTaskSlotsAfterRescale(
+                        evaluatedMetrics, scalingSummaries, numTaskSlotsUsed);
+        int newNumTms = (int) Math.ceil(numTaskSlotsAfterRescale / (double) taskSlotsPerTm);
+        LOG.info(
+                "Estimating {} task slots in use after rescale, spread across {} TaskManagers",
+                numTaskSlotsAfterRescale,
+                newNumTms);
+
+        MemorySize newHeapSize = new MemorySize((long) (usedTotalHeapSize / newNumTms));
+        // TM container memory can never grow beyond the user-specified max
+        Optional<MemorySize> maxMemory = context.getTaskManagerMemoryFromSpec();
+        if (maxMemory.isEmpty()) {
+            return Optional.empty();
+        }
+        // Apply limits
+        newHeapSize =
+                new MemorySize(
+                        Math.min(
+                                maxMemory.get().getBytes(),
+                                Math.max(
+                                        config.get(AutoScalerOptions.MEMORY_TUNING_MIN_HEAP)
+                                                .getBytes(),
+                                        newHeapSize.getBytes())));
```

Comment:
is it correct to compare here against `maxMemory`? `getTaskManagerMemoryFromSpec` returns the total process memory size, and only a small part of that may be heap.

Reply:
Furthermore the user may have configured a specific heap size directly in the conf

Reply:
True, this boundary is too high. We need to use the orignal max heap memory here. If users have configured a specific heap size, that is intended to be overwritten. But we could also choose to disable this feature in this case.